src/rocksdb/src/config.rs

```diff
@@ -13,11 +13,228 @@
+        self.parallelism = parallelism;
+        self.compression_type = compression_type;
+        self.bottommost_compression_type = bottommost_compression_type;
+        self.retry_max_duration = retry_max_duration;
+        self.stats_log_interval_seconds = stats_log_interval_seconds;
+        self.stats_persist_interval_seconds = stats_persist_interval_seconds;
+        self.point_lookup_block_cache_size_mb = point_lookup_block_cache_size_mb;
+
+        self.write_buffer_manager_config
+            .write_buffer_manager_memory_bytes = write_buffer_manager_memory_bytes;
+        self.write_buffer_manager_config
+            .write_buffer_manager_memory_fraction = write_buffer_manager_memory_fraction;
+        self.write_buffer_manager_config
+            .write_buffer_manager_allow_stall = write_buffer_manager_allow_stall;
+
+        // SeqCst is probably not required here, but its the easiest to reason about
+        self.dynamic.batch_size.store(batch_size, Ordering::SeqCst);
+    }
+}
+
+#[derive(Clone, Default)]
+pub struct SharedWriteBufferManager {
+    /// Keeping the shared [WriteBufferManager] behind an Arc and a Mutex.
+    /// The inner Arc's strong_count will be used to check if only the last
+    /// reference is remaining and then clean up the [WriteBufferManager].
+    shared: Arc<Mutex<Option<Arc<WriteBufferManager>>>>,
+}
+
+#[derive(Derivative)]
+#[derivative(Debug)]
+/// A handle to the [WriteBufferManager] which will be dropped when the
+/// `RocksDBInstance` will be dropped.
+pub(crate) struct WriteBufferManagerHandle {
+    #[derivative(Debug(format_with = "fmt_write_buffer_manager"))]
+    pub(crate) inner: Arc<WriteBufferManager>,
+    #[derivative(Debug = "ignore")]
+    manager: SharedWriteBufferManager,
+}
+
+fn fmt_write_buffer_manager(
+    buf: &Arc<WriteBufferManager>,
+    fmt: &mut std::fmt::Formatter,
+) -> Result<(), std::fmt::Error> {
+    fmt.debug_struct("WriteBufferManager")
+        .field("enabled", &buf.enabled())
+        .field("buffer_size", &buf.buffer_size())
+        .field("memory_usage", &buf.memory_usage())
+        .finish()
+}
+
+impl SharedWriteBufferManager {
+    pub(crate) fn get_or_init(&self, init_value: WriteBufferManager) -> WriteBufferManagerHandle {
+        let mut lock = self.shared.lock().expect("lock poisoned");
+        let wbm = Arc::<rocksdb::WriteBufferManager>::clone(
+            lock.get_or_insert_with(|| Arc::new(init_value)),
+        );
+        drop(lock);
+        WriteBufferManagerHandle {
+            inner: wbm,
+            manager: self.clone(),
+        }
+    }
+
+    pub(crate) fn get(&self) -> Option<Arc<WriteBufferManager>> {
+        let lock = self.shared.lock().expect("lock poisoned");
+        if let Some(buf) = lock.deref() {
+            Some(Arc::clone(buf))
+        } else {
+            None
+        }
+    }
+}
+
+/// Cleaning up the [`WriteBufferManager`] if there are only two reference counts
```

Comment:
As an alternative, is it possible to store a `Weak` at the central copy, so you wouldn't need this `Drop` implementation? This wouldn't work if you need the central strong reference.

Reply:
So, I was able to make the central copy a `Weak` and it did simplify the code. @guswynn wdyt?

Reply:
Added a test here https://github.com/MaterializeInc/materialize/pull/21340/commits/bfced2c3fd1b71217a90d57dee25dd9da06d40ca